gui/default/syncthing/core/syncthingController.js

```diff
@@ -3077,6 +3077,95 @@ angular.module('syncthing.core')
+                // Fallback for Internet Explorer.
+                window.clipboardData.setData('Text', text);
+                return alert(success);
+            } else if (document.queryCommandSupported) {
+                // Fallback for modern browsers on HTTP and non-IE old browsers.
+                // Check for document.queryCommandSupported("copy") support is
+                // omitted on purpose, as old Chrome versions reported "false"
+                // despite supporting the feature.
+                var textarea = document.createElement("textarea");
+                event.currentTarget.appendChild(textarea);
+                textarea.style.position = "fixed";
+                textarea.style.opacity = "0";
+                textarea.textContent = text;
+                textarea.select();
+                try {
+                    document.execCommand("copy");
+                    return alert(success);
+                }
+                catch (ex) {
+                    return alert(failure);
+                }
+                finally {
+                    event.currentTarget.removeChild(textarea);
+                }
+            }
+            return alert(failure);
+        };
+
+        $scope.shareDeviceId = function(method) {
+            // Title and footer can be reused between email, SMS, and possibly
+            // other methods, hence we define them separately in the beginning.
+            var title = $translate.instant("{%name%} has shared their Syncthing device ID!", {name: $scope.currentDevice.name});
+            var footer = $translate.instant("Learn more at {%address%}.", {address: "https://syncthing.net"});
+
+            switch (method) {
+                case 'email':
+                    var subject = title;
+                    var body = $translate.instant("You have been invited to connect with {%name%} using Syncthing!\r\n\r\nTo accept the invitation, add a new device using the ID below.", {name: $scope.currentDevice.name});
+                    body += '\r\n\r\n';
+                    body += $scope.currentDevice.deviceID;
+                    body += '\r\n\r\n';
+                    body += $translate.instant("Syncthing is a continuous file synchronization program. It synchronizes files between two or more computers in real time, safely protected from prying eyes. Your data is your data alone and you deserve to choose where it is stored, whether it is shared with some third party, and how it's transmitted over the internet.");
+                    body += '\r\n\r\n';
+                    body += footer;
+                    try {
+                        location.href = 'mailto:?subject=' + encodeURIComponent(subject) + '&body=' + encodeURIComponent(body);
+                        success = true;
+                    }
+                    catch (ex) {
+                        var error = $translate.instant("No email client found. Do you want to copy the email text to clipboard instead?");
```

Comment:
This message always appears for me (on Firefox), right before the mail client shows up.  Same with the SMS error message, where the desktop browser is still asking for an application to handle sms:// links.

Reply:
Yep, after more testing and investigation, it seems that detecting `mailto` and `sms` protocol support is simply impossible, especially when external applications are involved. I did find one very hacky method that measured browser response time and based its decision on that (i.e. it concluded no support for the protocol if the browser responded immediately), but I don't think it's the right way to go, as the browser may respond slowly e.g. on low-end hardware, etc. too.

I've switched to a completely different approach now, which asks the user for confirmation using the built-in modals instead of JS alerts.